week2/homework/src/todoList.js

```diff
@@ -0,0 +1,77 @@
+const fs = require('fs');
+const config = require('./config');
+
+//* WeakMap for private property(_todos) and methods(_read, _write)
+const _todos = new WeakMap();
+const _read = new WeakMap();
+const _write = new WeakMap();
+
+class TodoList {
+  constructor(file) {
+    this.file = file;
+
+    _read.set(this, file => {
+      const data = fs.readFileSync(file, { encoding: config.ENCODING });
+      return JSON.parse(data);
+    });
+
+    _write.set(this, data => {
+      fs.writeFileSync(this.file, JSON.stringify(data, null, 2));
+    });
+    _todos.set(this, _read.get(this)(this.file));
+  }
+
+  get todos() {
+    return _todos.get(this);
+  }
+
+  list() {
+    if (!this.todos.length) throw new Error('No Task');
+    return this.todos;
+  }
+
+  add(task) {
+    if (!task.length) {
+      throw new Error('Task required');
+    }
+    this.todos.push({
+      //* Id will be 1 if the todo list is empty, otherwise 1 greater than the max id
+      id: this.todos.length ? Math.max(...this.todos.map(i => i.id)) + 1 : 1,
+      description: task
+    });
+    _write.get(this)(this.todos);
+  }
+
+  remove(id) {
+    if (!this.todos.map(i => i.id).includes(parseInt(id))) {
+      throw new Error('Invalid Index');
+    }
+    const filteredTodos = this.todos
+      .filter(task => task.id !== id)
+      .map(task => task.description)
+      .reduce((tasks, description, index) => {
+        tasks.push({ id: index + 1, description });
```

Comment:
usually `id`s doesn't change, once it's on a one, it's just meant to be for that one.

imagine the following scenario when my browser knows it's id as `5` and the value of `301` for some reason and it's been increased or decreased for situation and use case, why should it use the next one if the previous client that uses this website having an event handler that fires on disconnect to remove it's counter, so I may start from `0` or I may not have a counter which in some case could be important

Reply:
I have tried to reorder the ids that is because `list` also shows the `id`s as indexes. You are right, the `id`s must be unique and should not be mutated also for data integrity and consistency. But in this homework i used the `id`s just for publishing the order of tasks in a sequential look.

I quit reordering the `id`s which also resulted a simpler code  😁